### PR TITLE
build-fhsenv-bubblewrap: fix fhsenv etc entries

### PR DESCRIPTION
**Pull request by hellwolf**
## Description of changes

This continues from #267504, which is related to #183874 and #279426.

* added more comments.
* symlink fhsenv etc entries when necessary.

This change may impact a lot of packages. I would like as for as many people reviewing as possible.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/build-support/build-fhsenv-bubblewrap/default.nix [hunk 1]
```diff
@@ -136,6 +138,7 @@ let
       fi
     done
 
+    # lopp through the entries of /etc in the fhs environment.
```
Comment (tejing1) [inline pkgs/build-support/build-fhsenv-bubblewrap/default.nix:141]:
```suggestion
    # loop through the entries of /etc in the fhs environment.
```

**Review by tejing1 — COMMENTED** [review]
Just from reading the code, it looks unobjectionable.

What I'm not entirely clear on is why this is necessary. What problem does it address?

**Comment by bjornfor** [comment]
> build-fhsenv-bubblewrap: fix fhsenv etc entries
> * added more comments.
> * symlink fhsenv etc entries when necessary.

This commit message does not explain enough for me. I don't understand what this fixes.

**Comment by hellwolf** [comment]
> Just from reading the code, it looks unobjectionable.
> 
> What I'm not entirely clear on is why this is necessary. What problem does it address?

From my understanding, when using `ro-bind` for symlinks (or perhaps regular files in general), it would result empty file in the fhs environment. See bug https://github.com/NixOS/nixpkgs/issues/258359 as an example.

**Comment by tejing1** [comment]
A quick test shows that `bwrap --ro-bind` binds the target of the symlink when used on a symlink, which shouldn't normally cause any problems.

In the case of `/etc/mtab` in the FHS env, it shows as zero length because the eventual target of the symlink chain is in `/proc`, and proc doesn't report file sizes accurately, but the contents are there.

```console
$ steam-run bash
$ ls -l /etc/mtab
-r--r--r-- 1 tejing users 0 Feb  4 17:27 /etc/mtab
$ wc -l /etc/mtab
71 /etc/mtab
```

I'm still not seeing a reason this is necessary.

**Comment by hellwolf** [comment]
> I'm still not seeing a reason this is necessary.

I am afraid at this point I will need to spend a bit more time to understand what's behind the scenes to answer you definitely.

**Comment by Atemu** [comment]
Does this fix https://github.com/NixOS/nixpkgs/issues/258359?

**Comment by hellwolf** [comment]
> Does this fix #258359?

It actually does. And I haven't done a complete analysis on why it actually fixed. It was @a-n-n-a-l-e-e who first brought to me that this is what I should look into further to fix the root cause.

**Comment by a-n-n-a-l-e-e** [comment]
> > Does this fix #258359?
> 
> It actually does. And I haven't done a complete analysis on why it actually fixed. It was @a-n-n-a-l-e-e who first brought to me that this is what I should look into further to fix the root cause.

i can't run the program in question but was just a guess given that /etc/mtab is a symlink outside of the container and the reported error seemed like a reasonable explanation.-- the wrapped runtime env should mirror the native runtime as close as possible and without the symlink it is not.

**Comment by tejing1** [comment]
I suppose if the program in question has special logic to deal with things in (or symlinks pointing to) `/proc`, then it would make sense that this change would fix the program's behavior.

**Comment by hellwolf** [comment]
Maybe this is an important detail:

By default, `insync start` spawns a detached demon. And a workaround I have have tried, without the mentioned fix in this PR, is to run `insync start --no-daemon`. By doing no-daemon, the problem went away.

I have just tried with replacing the last command with 'exec bash "$@"' in the run script and examined its /etc/mtab file:

```
$ trace-symlink /etc/mtab
/etc/mtab
/nix/store/xwkkxf0lg7wvfr4njcc6pvr3iv30abyq-insync-usr-target/etc/mtab
/nix/store/shwapdz34frds1hyp75sjgpwpq4g99s0-insync-chrootenv-etc/etc/mtab
/proc/mounts
/proc/self/mounts
```

**Comment by hellwolf** [comment]
I found another issue might be related: https://github.com/NixOS/nixpkgs/issues/224410

**Comment by tejing1** [comment]
No, that seems to be unrelated, based on the explanation given in the response.

**Comment by hellwolf** [comment]
I have an observation that makes me to think that without this fix, the current usage of `--ro-bind` to /etc/mtab is simply wrong.

I tested in a bwrap-ed bash session:

**without the fix, --ro-bind**

```
$ ps aux| grep --color -oE -- '--ro-bind [^ ]* /etc/mtab'
--ro-bind /nix/store/5k8qdzq6dclw2b9y9466cy0x4bwribvy-insync-fhs/etc/mtab /etc/mtab
$ ll /etc/mtab 
-r--r--r-- 1 hellwolf hellwolf 0 Mar  3 22:00 /etc/mtab
$ mount | grep mtab
proc on /etc/mtab type proc (ro,nosuid,nodev,noexec,relatime)
```

This doesn't make any sense to me, because I think the mtab is bound to a specific file under `/proc/???/mounts`.

**with the fix, --symlink**

```
$ ps aux| grep --color -oE -- '--symlink [^ ]* /etc/mtab'
--symlink /nix/store/5k8qdzq6dclw2b9y9466cy0x4bwribvy-insync-fhs/etc/mtab /etc/mtab
$ ll /etc/mtab 
lrwxrwxrwx 1 hellwolf hellwolf 63 Mar  3 22:05 /etc/mtab -> /nix/store/5k8qdzq6dclw2b9y9466cy0x4bwribvy-insync-fhs/etc/mtab
```

This looks more sane.

Here is a sample of the long commands when using the fix:

```
hellwolf  591409  0.0  0.0   3468  1920 pts/1    S    22:05   0:00 /nix/store/w4mrg85mlqjsiz19jzc4i1xnb7x92rad-bubblewrap-0.8.0/bin/bwrap --dev-bind /dev /dev --proc /proc --chdir /home/hellwolf/Projects/NixOS/nixpkgs.workspace/insync-mtab-bug --die-with-parent --ro-bind /nix /nix --tmpfs /nix/store/8mc30d49ghc8m5z96yz39srlhg5s9sjj-glibc-2.38-44/etc --tmpfs /etc --symlink /etc/ld.so.conf /nix/store/8mc30d49ghc8m5z96yz39srlhg5s9sjj-glibc-2.38-44/etc/ld.so.conf --symlink /etc/ld.so.cache /nix/store/8mc30d49ghc8m5z96yz39srlhg5s9sjj-glibc-2.38-44/etc/ld.so.cache --ro-bind /nix/store/8mc30d49ghc8m5z96yz39srlhg5s9sjj-glibc-2.38-44/etc/rpc /nix/store/8mc30d49ghc8m5z96yz39srlhg5s9sjj-glibc-2.38-44/etc/rpc --remount-ro /nix/store/8mc30d49ghc8m5z96yz39srlhg5s9sjj-glibc-2.38-44/etc --tmpfs /nix/store/x999l88j5rw8rfx9c16ba5jfrrw4l24k-glibc-2.38-44/etc --symlink /etc/ld.so.conf /nix/store/x999l88j5rw8rfx9c16ba5jfrrw4l24k-glibc-2.38-44/etc/ld.so.conf --symlink /etc/ld.so.cache /nix/store/x999l88j5rw8rfx9c16ba5jfrrw4l24k-glibc-2.38-44/etc/ld.so.cache --ro-bind /nix/store/x999l88j5rw8rfx9c16ba5jfrrw4l24k-glibc-2.38-44/etc/rpc /nix/store/x999l88j5rw8rfx9c16ba5jfrrw4l24k-glibc-2.38-44/etc/rpc --remount-ro /nix/store/x999l88j5rw8rfx9c16ba5jfrrw4l24k-glibc-2.38-44/etc --ro-bind /nix/store/5k8qdzq6dclw2b9y9466cy0x4bwribvy-insync-fhs/usr /usr --ro-bind /etc /.host-etc --symlink usr/bin /bin --symlink usr/lib /lib --symlink usr/lib64 /lib64 --symlink usr/sbin /sbin --symlink /nix/store/5k8qdzq6dclw2b9y9466cy0x4bwribvy-insync-fhs/etc/login.defs /etc/login.defs --symlink /nix/store/5k8qdzq6dclw2b9y9466cy0x4bwribvy-insync-fhs/etc/mtab /etc/mtab --symlink /nix/store/5k8qdzq6dclw2b9y9466cy0x4bwribvy-insync-fhs/etc/pam.d /etc/pam.d --symlink /nix/store/5k8qdzq6dclw2b9y9466cy0x4bwribvy-insync-fhs/etc/profile /etc/profile --symlink /nix/store/5k8qdzq6dclw2b9y9466cy0x4bwribvy-insync-fhs/etc/profile.d /etc/profile.d --symlink /nix/store/5k8qdzq6dclw2b9y9466cy0x4bwribvy-insync-fhs/etc/rpc /etc/rpc --symlink /.host-etc/static /etc/static --symlink /.host-etc/nix /etc/nix --symlink /.host-etc/shells /etc/shells --symlink /.host-etc/bashrc /etc/bashrc --symlink /.host-etc/passwd /etc/passwd --symlink /.host-etc/group /etc/group --symlink /.host-etc/shadow /etc/shadow --symlink /.host-etc/hosts /etc/hosts --symlink /.host-etc/resolv.conf /etc/resolv.conf --symlink /.host-etc/nsswitch.conf /etc/nsswitch.conf --symlink /.host-etc/profiles /etc/profiles --symlink /.host-etc/sudoers /etc/sudoers --symlink /.host-etc/localtime /etc/localtime --symlink /.host-etc/zoneinfo /etc/zoneinfo --symlink /.host-etc/machine-id /etc/machine-id --symlink /.host-etc/os-release /etc/os-release --symlink /.host-etc/fonts /etc/fonts --symlink /.host-etc/alsa /etc/alsa --symlink /.host-etc/ssl/certs /etc/ssl/certs --symlink /.host-etc/pki /etc/pki --bind /boot /boot --bind /home /home --bind /lost+found /lost+found --bind /mnt /mnt --bind /opt /opt --bind /root /root --bind /run /run --bind /srv /srv --bind /sys /sys --bind /tmp /tmp --bind /var /var --tmpfs /tmp/.X11-unix --ro-bind-try /tmp/.X11-unix/X0 /tmp/.X11-unix/X0 /nix/store/khzgfszn0vqwkgw2110n3wkp04ymmnq3-insync-init
```

**Comment by tejing1** [comment]
That's just how bind mounts get reported by `mount`. The "source" of any bind mount is just the source of the original file system it's bound from, and the source of the proc filesystem is set to the dummy value `proc`.

If you bind mount `/foo/bar` onto `/bar`, and `/foo` is mounted from `/dev/sda1`, then mount will show an entry like `/dev/sda1 on /bar type <whatever>`.

So in short, working as expected for a bind mount.

**Comment by hellwolf** [comment]
> So in short, working as expected for a bind mount.

Yes.

But not working for what is needed, hence the fix was needed.

Because if you bind a `/proc/some_others_pid/mounts`, then that is not what you need which should be `/proc/self/mounts`.

I may still not be able to write a longer analysis for why exactly it is broken for some. But the fact is, it  doesn't work for the end app without the fix.

**Comment by tejing1** [comment]
Ok, your comment is a little hard to parse, but I do see one actual issue caused by the current behavior, now that you point it out.

The bind mount essentially dereferences `/proc/self` once and locks it in, making other processes get the `/proc/$$/mounts` for the wrong PID when they read `/etc/mtab`. That shouldn't matter in most cases, but if you enter yet another layer of user namespaces, *and* that new layer isn't another instance of `buildFHSEnv`, then it starts returning the wrong data.

I'm not at all convinced that this is the cause of the problem you're actually having, but it's a substantive argument for why it should be changed, at least. I'm sorry if it seems like I'm dragging things down here, I just want to make sure we're implementing the *right* fix, rather than just *some* fix, and I still don't understand the problem well enough to know what the right fix is.

**Comment by hellwolf** [comment]
I will find some time to analyze the problem a bit more some other time, and write better description.

Though, as you see, the current behavior is wrong, and the fix at least doesn't make it worse. We could either spend more time on understanding a 3rd-party software, or we push this fix out for now and notify the downstream affected packages. But in the former case, it would be good to timebox it too.

**Comment by hellwolf** [comment]
Here is the summary of the problem statement:

The "insync" package is third-party non-free software written in PySide and shipped as Ubuntu/Debian/Linux Mint/Fedora packages. Its nixpkgs package uses the `build-fhsenv`. Its latest PR is #279426.

Since late last year, the package has started emitting a sounding error: "Invalid argument: /etc/mtab." However, due to the package's proprietary nature, it is hard to debug. I have asked this in the official forum, too; it was not clear it would ever be resolved since support may see it as a non-supported Linux issue.

After several rounds of workarounds and investigations, a community member pointed out that it may have been related to the `build-fhsenv` after all. Hence, this PR was introduced. The test also shows that this PR will indeed remove the "insync" package errors.

Again, since "insync" is a 3rd party unfree proprietary software, making further investigation into why and how it needs to read "/etc/mtab" file may be futile and unnecessary. But it suffices to say that /etc/mtab file, by design, should be a symlink that points to `/proc/self/mounts` where `/etc/self ` is a Linux peculiarity that "bwrap bind" would break. That is likely because, "bwrap bind", when dealing with a file, will bind the canonicalize path (similar to `readlink -f`) of the original file to the destination. 

----

I hope this explanation is conclusive, I really would like us move forward and merge this fix. So that people using insync could be unblocked!

File: pkgs/build-support/build-fhsenv-bubblewrap/default.nix (hunk 1 above)
Comment (tejing1) [inline pkgs/build-support/build-fhsenv-bubblewrap/default.nix:141]:
```suggestion
    # loop through the entries of /etc in the fhs environment.
```
Somehow this got un-fixed after we discussed it previously.

**Review by tejing1 — CHANGES_REQUESTED** [review]

File: pkgs/build-support/build-fhsenv-bubblewrap/default.nix (hunk 1 above)
Comment (hellwolf) [reply pkgs/build-support/build-fhsenv-bubblewrap/default.nix:141]: done!

**Review by tejing1 — APPROVED** [review]

**Comment by hellwolf** [comment]
let me do a `nixpkgs-review pr 286360 --print-result`.

**Comment by Atemu** [comment]
I will take a closer look at this again next week. Please ping me in case I forget.

**Comment by hellwolf** [comment]
> I will take a closer look at this again next week. Please ping me in case I forget.

Hi, any chance you could have a look this week?

**Comment by hellwolf** [comment]
Good to hear!

Shall we move this forward?

**Comment by hellwolf** [comment]
CC @Atemu

**Review by ghost — APPROVED** [review]
this has been languishing for months.
